client/src/main/java/com/vesoft/nebula/storage/client/StorageClientImpl.java

```diff
@@ -191,57 +139,86 @@ public boolean put(int part, String key, String value) {
-                        updateLeader(space, part, currentLeaderAddress);
-                    }
-                    return true;
-                }
-            } catch (TException e) {
-                LOGGER.error(String.format("Put Failed: %s", e.getMessage()));
-            }
-        }
-        return false;
+        return doPut(space, client, request);
     }
 
     /**
      * Put multi key-value pairs into partition
      *
-     * @param part   partitionID
-     * @param values key-value pairs
+     * @param space nebula space id
+     * @param kvs   key-value pairs
      * @return
      */
     @Override
-    public boolean put(int part, Map<String, String> values) {
-        checkLeader(part);
-        PutRequest request = new PutRequest();
-        request.setSpace_id(space);
-        Map<Integer, List<Pair>> parts = Maps.newHashMap();
-        List<Pair> pairs = Lists.newLinkedList();
-        for (Map.Entry<String, String> entry : values.entrySet()) {
-            pairs.add(new Pair(entry.getKey(), entry.getValue()));
+    public boolean put(int space, Map<String, String> kvs) {
+        Map<Integer, List<Pair>> groups = new HashMap<>();
+        for (Map.Entry<String, String> kv : kvs.entrySet()) {
+            int part = keyToPartId(space, kv.getKey());
+            if (!groups.containsKey(part)) {
+                groups.put(part, new ArrayList<Pair>());
+            }
+            groups.get(part).add(new Pair(kv.getKey(), kv.getValue()));
         }
-        parts.put(part, pairs);
-        request.setParts(parts);
-        LOGGER.debug(String.format("Put Request: %s", request.toString()));
 
+        Map<HostAddr, PutRequest> requests = new HashMap<>();
+        for (Map.Entry<Integer, List<Pair>> entry : groups.entrySet()) {
+            int part = entry.getKey();
+            HostAddr leader = getLeader(space, part);
+            if (!requests.containsKey(leader)) {
+                PutRequest request = new PutRequest();
+                request.setSpace_id(space);
+                Map<Integer, List<Pair>> parts = Maps.newHashMap();
+                parts.put(part, entry.getValue());
+                request.setParts(parts);
+                LOGGER.debug(String.format("Put Request: %s", request.toString()));
+                requests.put(leader, request);
+            } else {
+                PutRequest request = requests.get(leader);
+                if (!request.parts.containsKey(part)) {
+                    request.parts.put(part, entry.getValue());
+                } else {
+                    request.parts.get(part).addAll(entry.getValue());
+                }
+            }
+        }
+
+        final CountDownLatch countDownLatch = new CountDownLatch(groups.size());
+        final List<Boolean> responses = Collections.synchronizedList(new ArrayList<Boolean>(groups.size()));
+        ExecutorService executorService = Executors.newFixedThreadPool(groups.size());
```

Comment:
why don't move `executorService` into class instance scope ? `executorService` maybe a heavily object, create and release frequently would influence performance.

Reply:
:thumb: